rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+    inference graph to initialization graph
+
+<img src="20190722-tflite-training/trans2.png" width="600" />
+
+### Control Flow
+
+Control flow is not required for basic training. However it's useful in the
+following cases:
+
+*   **Train models with control flow**: For a model with control flow, control
+    flow ops are required for both inference and training. E.g. It requires `If`
+    to compute the gradient of `If`, and it requires `While` loop to compute the
+    gradient of `While`.
+*   **Run training loops** inside TensorFlow Lite models: To train a model, it's
+    usually required to invoke the training step multiple times. It's doable by
+    writing a loop in application code to invoke the training subgraph multiple
+    times. It can also be implemented by using a `While` op to iterate over the
+    training data.
+
+The detailed design of supporting control flow in TensorFlow Lite is at
+[This RFC](https://github.com/tensorflow/community/pull/83).
+
+### Tensor List
+
+Tensor List is not required for basic training. However it's important for
+training models with control flow loops.
+
+Whenever there are loops (e.g. While) in the inference graph, Tensor List is
+usually required to train the model. Even if Tensor List ops aren't in the
+original inference graph, when constructing the training graph, Tensor List ops
+are often added automatically to memorize intermediate values, for computing
+gradients over the loop.
+
+Tensor Lists can also be used without loops but this rarely happens. In this
+case, Tensor List ops are required for inference and training.
+
+### Saving Format
+
+We don't need a saving format to make training work. The newly trained weights
+can be used in TensorFlow Lite directly for inference. The weights data are
+stored in (variable) tensors, and users can read and write these data through
+interpreter API. This already enables saving and loading trainable weights from
+the application code. On top of this, we can define a saving format to store
+trainable weights.
+
+**(Open question)** There are a few potential options:
+
+*   Option 0: Do nothing
+*   Option 1: Implement TensorFlow Lite's own variable saving file format. This
+    can be a simple binary format that stores variable ID -> data mapping.
+*   Option 2: Write the variable value back to TensorFlow Lite model.
+    *   Pros: Only having a single file & single file format.
+    *   Cons: Have to ship FlatBuffer writer in the binary. Have to write the
+        entire model (which may contain other huge frozen constants) back to
+        storage.
+*   Option 3: Use TensorFlow checkpoint format
+    *   3A: Implement checkpoint reading / writing functionality in TensorFlow
+        Lite
+    *   3B: Reading / writing through Flex runtime (so it only works when Flex
+        runtime is used)
+
+### Feeding data with tf.Example / tf.DataSet
+
+`tf.Example` and `tf.DataSet` are technically not training-specific
```

Comment:
nit: tf.data.Dataset is I presume what you mean?

Reply:
Yes (leaving this open as a TODO)

Reply:
Fixed